Doc/library/os.rst

```diff
@@ -2366,7 +2366,8 @@ features:
 
    .. attribute:: st_ino
 
-      Inode number.
+      Inode number. On Windows, this is an arbitrary value that uniquely
+      identifies the file on its device.
```

Comment:
Now you put it that way, that's what an inode number is on UNIX too. :-)

Is it really not useful to explain how the arbitrary value is assembled?

Reply:
`st_ino` is from the file system `IndexNumber` in the [`FileInternalInformation`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/ns-ntifs-_file_internal_information), and `st_dev` is from the `VolumeSerialNumber` in the [`FileFsVolumeInformation`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntddk/ns-ntddk-_file_fs_volume_information). The VSN is not necessarily unique on the system, and neither value is necessarily stable or non-zero. Here are the definitions from File System Control Codes [MS-FSCC].

[2.4.20 FileInternalInformation](https://msdn.microsoft.com/en-us/library/cc232072.aspx)
> IndexNumber (8 bytes): A 64-bit signed integer that contains the 8-byte file reference number for the file. This number MUST be assigned by the file system and is unique to the volume on which the file or directory is located. This file reference number is the same as the file reference number that is stored in the FileId field of the FILE_ID_BOTH_DIR_INFORMATION and FILE_ID_FULL_DIR_INFORMATION data elements.<110> The value of this field MUST be greater than or equal to 0. For file systems which do not support a file reference number, this field MUST be set to 0, and MUST be ignored.

[2.5.9 FileFsVolumeInformation](https://msdn.microsoft.com/en-us/library/cc232108.aspx)
> VolumeSerialNumber (4 bytes): A 32-bit unsigned integer that contains the serial number of the volume. The serial number is an opaque value generated by the file system at format time, and is not necessarily related to any hardware serial number for the device on which the file system is located. No specific format or content of this field is required for protocol interoperation. This value is not required to be unique.

These values are generally good enough, in particular with common local file systems on Windows (NTFS, ReFS, FAT32, exFAT, UDF). `IndexNumber` can change on FAT32 drives due to 3rd party defrag tools, since it's based on the file's byte offset in the directory entry. Microsoft's built-in defrag is careful to not change this.

UNC providers may not provide useful values here. For example, a WebDAV network drive returns 0 for both values , which breaks `os.path.samefile`.

Reply:
> Now you put it that way, that's what an inode number is on UNIX too. :-)

Fairly sure that's why I never documented it in the first place :)

I don't believe it's useful to document the exact way it is created, simply because it is meant to be opaque. I suspect it's more opaque on Windows than UNIX, as there is literally nothing useful you can do except compare it to another value from the same device, and even then it may not mean anything.

You can't get from the number back to the file, and it's only "assembled" because it is a 64-bit value passed as two 32-bit values in a struct, which is common for Windows APIs that have been around for a long time.

Reply:
It would still be nice to be able to link this back to "authoritative" docs from MS. On UNIX, st_ino is well-known. But you're giving me nothing I can search for on Windows. If you had mentioned this was formed by combining nFileIndexHigh/Low, I could search the MS docs. So I still think that there is something useful (I admit that the specific way the High/Low parts are combined is uninteresting as long as we know it's not via xor :-).

Reply:
No, the file index number is useful. If the file system assigns a non-zero index value to a file, then you can use it with [`OpenFileById`](https://msdn.microsoft.com/en-us/library/aa365432). What's not generally useful is the VSN, which we're using for `st_dev`. There's no equivalent to POSIX `st_dev` in Windows, which is why `OpenFileById` requires an existing handle opened for a file on the volume or share.

Reply:
In case of any doubt that this `st_ino` value can be used with `OpenFileById`:

    >>> open(r'C:\Temp\spam.txt', 'w').write('eggs')
    4
    >>> s = os.stat(r'C:\Temp\spam.txt')
    >>> fdvol = os.open(r'\\?\C:', 0)
    >>> hvol = msvcrt.get_osfhandle(fdvol)
    >>> hfile = win32file.OpenFileById(hvol, s.st_ino, win32file.GENERIC_READ, 7, 0)
    >>> win32file.ReadFile(hfile, 4)
    (0, b'eggs')